zapcore/buffered_write_syncer.go

```diff
@@ -0,0 +1,147 @@
+type BufferedWriteSyncer struct {
+	WriteSyncer
+
+	// Size specifies the maximum amount of data the writer will buffer before
+	// flushing. Defaults to 256 kB.
+	Size int
+	// FlushInterval specifies how often the writer should flush data if there
+	// have been no writes. Defaults to 30 seconds.
+	FlushInterval time.Duration
+	// Clock, if specified, provides control of the source of time for the
+	// writer. Uses the system clock by default.
+	Clock Clock
+
+	// unexported fields for state
+	ws          WriteSyncer
+	mu          sync.Mutex
+	writer      *bufio.Writer
+	ticker      *time.Ticker
+	stop        chan struct{}
+	initialized bool
+}
+
+const (
+	// _defaultBufferSize specifies the default size used by Buffer.
+	_defaultBufferSize = 256 * 1024 // 256 kB
+
+	// _defaultFlushInterval specifies the default flush interval for
+	// Buffer.
+	_defaultFlushInterval = 30 * time.Second
+)
+
+func (s *BufferedWriteSyncer) initialize() {
+	s.ws = s.WriteSyncer
+
+	size := s.Size
+	if size == 0 {
+		size = _defaultBufferSize
+	}
+
+	flushInterval := s.FlushInterval
+	if flushInterval == 0 {
+		flushInterval = _defaultFlushInterval
+	}
+
+	if s.Clock == nil {
+		s.Clock = DefaultClock
+	}
+	s.ticker = s.Clock.NewTicker(flushInterval)
+
+	writer := s.WriteSyncer
+	if w, ok := writer.(*lockedWriteSyncer); ok {
+		writer = w.ws
+	} // don't double lock
+	s.writer = bufio.NewWriterSize(writer, size)
+
+	s.stop = make(chan struct{})
+	s.initialized = true
+	go s.flushLoop()
+}
+
+// Write writes log data into buffer syncer directly, multiple Write calls will be batched,
+// and log data will be flushed to disk when the buffer is full or periodically.
+func (s *BufferedWriteSyncer) Write(bs []byte) (int, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.initialized {
+		s.initialize()
+	}
+
+	// To avoid partial writes from being flushed, we manually flush the existing buffer if:
+	// * The current write doesn't fit into the buffer fully, and
+	// * The buffer is not empty (since bufio will not split large writes when the buffer is empty)
+	if len(bs) > s.writer.Available() && s.writer.Buffered() > 0 {
+		if err := s.writer.Flush(); err != nil {
+			return 0, err
+		}
+	}
+
+	return s.writer.Write(bs)
+}
+
+// Sync flushes buffered log data into disk directly.
+func (s *BufferedWriteSyncer) Sync() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return multierr.Append(s.writer.Flush(), s.ws.Sync())
+}
+
+// flushLoop flushes the buffer at the configured interval until Close is
+// called.
+func (s *BufferedWriteSyncer) flushLoop() {
+	for {
+		select {
+		case <-s.ticker.C:
+			// we just simply ignore error here
+			// because the underlying bufio writer stores any errors
+			// and we return any error from Sync() as part of the close
+			_ = s.Sync()
+		case <-s.stop:
+			return
+		}
+	}
+}
+
+// Close closes the buffer, cleans up background goroutines, and flushes
+// remaining, unwritten data. This will not close the underlying WriteSyncer.
```

Comment:
The warning that this will not close the underlying `WriteSyncer` makes me wonder if `Close` is a little surprising. Would this be better named something else like `Stop` to make it very clear that it doesn't close the underlying writer?

Reply:
Reasonable. Renaming.

Reply:
Sounds good. `Stop` makes more sense in the context of this specific writer.